File: kotlin/decorator.kt
interface Coffee {
    fun cost(): Int
}

class SimpleCoffee : Coffee {
    override fun cost(): Int = 5
}

open class CoffeeDecorator(private val coffee: Coffee) : Coffee {
    override fun cost(): Int = coffee.cost()
}

class MilkDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost(): Int = super.cost() + 2
}

class SugarDecorator(coffee: Coffee) : CoffeeDecorator(coffee) {
    override fun cost(): Int = super.cost() + 1
}

fun main() {
    val coffee: Coffee = SimpleCoffee()
    println(coffee.cost()) // 5

    val coffeeWithMilk: Coffee = MilkDecorator(coffee)
    println(coffeeWithMilk.cost()) // 7

    val coffeeWithSugar: Coffee = SugarDecorator(coffee)
    println(coffeeWithSugar.cost()) // 6

    val coffeeWithMilkAndSugar: Coffee = SugarDecorator(coffeeWithMilk)
    println(coffeeWithMilkAndSugar.cost()) // 8
}

File: cpp/decorator.cpp
#include <iostream>
using namespace std;

class Coffee {
public:
    virtual int cost() const {
        return 5;
    }
    virtual ~Coffee() = default;
};

class CoffeeDecorator : public Coffee {
protected:
    Coffee* coffee;
public:
    CoffeeDecorator(Coffee* coffee) : coffee(coffee) {}
    virtual int cost() const override {
        return coffee->cost();
    }
};

class MilkDecorator : public CoffeeDecorator {
public:
    MilkDecorator(Coffee* coffee) : CoffeeDecorator(coffee) {}
    int cost() const override {
        return coffee->cost() + 2;
    }
};

class SugarDecorator : public CoffeeDecorator {
public:
    SugarDecorator(Coffee* coffee) : CoffeeDecorator(coffee) {}
    int cost() const override {
        return coffee->cost() + 1;
    }
};

int main() {
    Coffee* coffee = new Coffee();
    cout << coffee->cost() << endl; // 5

    Coffee* coffeeWithMilk = new MilkDecorator(coffee);
    cout << coffeeWithMilk->cost() << endl; // 7

    Coffee* coffeeWithSugar = new SugarDecorator(coffee);
    cout << coffeeWithSugar->cost() << endl; // 6

    Coffee* coffeeWithMilkAndSugar = new SugarDecorator(coffeeWithMilk);
    cout << coffeeWithMilkAndSugar->cost() << endl; // 8

    // 메모리 해제
    delete coffeeWithMilkAndSugar;
    delete coffeeWithSugar;
    delete coffeeWithMilk;
    delete coffee;
}
